runtime/lua/man.lua

```diff
@@ -152,21 +220,529 @@ local function highlight_line(line, linenr)
+  -- http://comments.gmane.org/gmane.editors.vim.devel/29085
+  -- Set MAN_KEEP_FORMATTING so Debian man doesn't discard backspaces.
+  local cmd = { 'env', 'MANPAGER=cat', 'MANWIDTH=' .. manwidth, 'MAN_KEEP_FORMATTING=1', 'man' }
+  if localfile_arg then
+    cmd[#cmd + 1] = '-l'
+  end
+  cmd[#cmd + 1] = path
+  return man_system(cmd, silent)
+end
+
+local function put_page(page)
+  vim.bo.modified = true
+  vim.bo.readonly = false
+  vim.bo.swapfile = false
+
+  api.nvim_buf_set_lines(0, 0, -1, false, vim.split(page, '\n'))
+
+  while fn.getline(1):match('^%s*$') do
+    api.nvim_buf_set_lines(0, 0, 1, false, {})
+  end
+  -- XXX: nroff justifies text by filling it with whitespace.  That interacts
+  -- badly with our use of $MANWIDTH=999.  Hack around this by using a fixed
+  -- size for those whitespace regions.
+  vim.cmd([[silent! keeppatterns keepjumps %s/\s\{199,}/\=repeat(' ', 10)/g]])
+  vim.cmd('1') -- Move cursor to first line
+  highlight_man_page()
+  set_options(false)
+end
+
+local function format_candidate(path, psect)
+  if matchstr(path, [[\.\%(pdf\|in\)$]]) then -- invalid extensions
+    return ''
+  end
+  local sect, name = extract_sect_and_name_path(path)
+  if sect == psect then
+    return name
+  elseif sect and name and matchstr(sect, psect .. '.\\+$') then -- invalid extensions
+    -- We include the section if the user provided section is a prefix
+    -- of the actual section.
+    return ('%s(%s)'):format(name, sect)
+  end
+  return ''
+end
+
+local function get_paths(sect, name, do_fallback)
+  -- callers must try-catch this, as some `man` implementations don't support `s:find_arg`
+  local ok, ret = pcall(function()
+    local mandirs = table.concat(vim.split(man_system({ 'man', find_arg }), '[:\n]'), ',')
+    local paths = fn.globpath(mandirs, 'man?/' .. name .. '*.' .. sect .. '*', false, true)
+    pcall(function()
+      -- Prioritize the result from verify_exists as it obeys b:man_default_sects.
+      local first = verify_exists(sect, name)
+      paths = vim.tbl_filter(function(v)
+        return v ~= first
+      end, paths)
+      paths = { first, unpack(paths) }
+    end)
+    return paths
+  end)
+
+  if not ok then
+    if not do_fallback then
+      error(ret)
+    end
+
+    -- Fallback to a single path, with the page we're trying to find.
+    ok, ret = pcall(verify_exists, sect, name)
+
+    return { ok and ret or nil }
+  end
+  return ret or {}
+end
+
+local function complete(sect, psect, name)
+  local pages = get_paths(sect, name, false)
+  -- We remove duplicates in case the same manpage in different languages was found.
+  return fn.uniq(fn.sort(vim.tbl_map(function(v)
+    return format_candidate(v, psect)
+  end, pages) or {}, 'i'))
+end
+
+-- see extract_sect_and_name_ref on why tolower(sect)
+function M.man_complete(arg_lead, cmd_line, _)
+  -- TODO(lewis6991): for some reason using vim.fn.split here makes completion very laggy
+  local args = vim.split(cmd_line, '%s+')
```

Comment:
To equivalent to `vim.fn.split()`, `trimempty = true`. It make a difference here. Not sure about other places, but I guess it is better to add `trimempty = true` to other places as well.
```suggestion
  local args = vim.split(cmd_line, '%s+', {trimempty = true})
```

Reply:
Currently completion results contain empty strings. When I tried adding `trimempty = true` to every `vim.split()` empty strings no longer appear.

Reply:
Ahhh! This explains why the old implementation was really laggy. If there was no arg_lead then `args == {'Man'}` which caused `complete('', '', '')` to be called which is really slow. Without `trimempty` then the `#args==1` branch is unreachable.